homeassistant/components/doorbird/__init__.py

```diff
@@ -59,55 +59,61 @@ def setup(hass, config):
-                str(status[1]),
-            )
-            return False
-
-        # Subscribe to doorbell or motion events
-        if events:
-            try:
-                doorstation.register_events(hass)
-            except HTTPError:
-                hass.components.persistent_notification.create(
-                    "Doorbird configuration failed.  Please verify that API "
-                    "Operator permission is enabled for the Doorbird user. "
-                    "A restart will be required once permissions have been "
-                    "verified.",
-                    title="Doorbird Configuration Failure",
-                    notification_id="doorbird_schedule_error",
+        try:
+            device_ip = doorstation_config.get(CONF_HOST)
+            username = doorstation_config.get(CONF_USERNAME)
+            password = doorstation_config.get(CONF_PASSWORD)
+            custom_url = doorstation_config.get(CONF_CUSTOM_URL)
+            events = doorstation_config.get(CONF_EVENTS)
+            token = doorstation_config.get(CONF_TOKEN)
+            name = doorstation_config.get(CONF_NAME) or "DoorBird {}".format(index + 1)
+
+            device = DoorBird(device_ip, username, password)
+            status = device.ready()
+
+            if status[0]:
+                doorstation = ConfiguredDoorBird(
+                    device, name, events, custom_url, token
+                )
+                doorstations.append(doorstation)
+                _LOGGER.info(
+                    'Connected to DoorBird "%s" as %s@%s',
+                    doorstation.name,
+                    username,
+                    device_ip,
+                )
+            elif status[1] == 401:
+                _LOGGER.error(
+                    "Authorization rejected by DoorBird for %s@%s", username, device_ip
                 )
-
                 return False
+            else:
+                _LOGGER.error(
+                    "Could not connect to DoorBird as %s@%s: Error %s",
+                    username,
+                    device_ip,
+                    str(status[1]),
+                )
+                return False
+
+            # Subscribe to doorbell or motion events
+            if events:
+                try:
+                    doorstation.register_events(hass)
+                except HTTPError:
+                    hass.components.persistent_notification.create(
+                        "Doorbird configuration failed.  Please verify that API "
+                        "Operator permission is enabled for the Doorbird user. "
+                        "A restart will be required once permissions have been "
+                        "verified.",
+                        title="Doorbird Configuration Failure",
+                        notification_id="doorbird_schedule_error",
+                    )
+
+                    return False
+        # pylint: disable=broad-except
+        except Exception as error:
```

Comment:
Do not disable broad-except instead catch only the exceptions that can occur

Reply:
Candidly, I'd probably give exactly the same feedback on a quick review -- see my commit message ("I don't disable a lint warning lightly, but this is the entire intent of this PR") to know that I was thoughtful about this when doing it. That exception is disabled 80+ times in Home Assistant, and a broad exception is allowed in two cases per PEP8, and this is exactly case 1:

1. If the exception handler will be printing out or logging the traceback; at least the user will be aware that an error has occurred.

(N.B., I've just improved the code to use _LOGGER.exception instead of .error so that we get a traceback.)

That address the broad except, and the reason why I believe it's better to include the entire iteration code in the try block is that that is 100% the intent of the PR and it makes confirming that the code is correct towards that intent trivially: one iteration of the loop cannot make the package fail and that is what we're trying to avoid.

The alternative of reasoning carefully about which code segments can cause exceptions (and potentially which exceptions) in a language that does not put throws declarations in the interface of functions is not a wise investment because: 

1) it requires deep analysis of the code to ensure all exceptions are caught;
2) it is fragile if the underlying package or support module throw new exceptions; and
3) it results in redundant try/except blocks that are going to do the same thing (log the exception, show a stack trace, and proceed with the next iteration of the loop)

My intent down the road is to support better retrying of a failed doorbird device and that will require some of that deeper analysis, but the current code doesn't reflect the reality that doorbird devices may fail independently and should not bring down the whole subsystem with a single failure.

Thanks!

Reply:
I am so sorry I did not read that commit message at all. 😅
Just glanced over the rest and thought that this would be a good point for a review. Completely agree with your reasoning and yes it is very important to at least log the traceback in case of an exception.
I will close this review now ✌

Reply:
Thanks -- like I wrote, I agree with your point in the general case. 

BTW, do you know what am I supposed to do about the tests that are failing due to an infrastructure bug (some problem with virtual envs that has nothing to do with my code)?

Reply:
No, I have submitted a couple of PRs as well and they all failed. Not really sure what caused this problem or how to fix it at the moment.

Reply:
We should not put a try…except around the whole block. There are specific lines that will raise, wrap those and `continue` if it fails. 

We should only catch `Exception` if we are running passed in callbacks and don't know what to expect. Since the API lib is open source, we know exactly what to expect.